lib/hanami/repository.rb

```diff
@@ -118,769 +106,322 @@ module Hanami
-      #       query do
-      #         where(published: true)
-      #       end
-      #     end
-      #
-      #     def self.rank
-      #       # reuse .published, which returns a query that respond to #desc
-      #       published.desc(:comments_count)
-      #     end
-      #
-      #     def self.best_article_ever
-      #       # reuse .published, which returns a query that respond to #limit
-      #       rank.limit(1)
-      #     end
-      #
-      #     def self.comments_average
-      #       query.average(:comments_count)
-      #     end
-      #   end
-      def query(&blk)
-        adapter.query(collection, self, &blk)
+      #   repository = UserRepository.new
+      #   user       = repository.create(name: 'Luca')
+      #
+      #   user       = repository.delete(user.id)
+      def delete(*args)
+        super
+      rescue => e
+        raise Hanami::Model::Error.for(e)
       end
+    end
 
-      # Negates the filtering conditions of a given query with the logical
-      # opposite operator.
-      #
-      # This is only supported by the SqlAdapter.
-      #
-      # @param query [Object] a query
-      #
-      # @return a negated query, the type depends on the current adapter
-      #
-      # @api public
-      # @since 0.1.0
-      #
-      # @see Hanami::Model::Adapters::Sql::Query#negate!
-      #
-      # @example
-      #   require 'hanami/model'
-      #
-      #   class ProjectRepository
-      #     include Hanami::Repository
-      #
-      #     def self.cool
-      #       query do
-      #         where(language: 'ruby')
-      #       end
-      #     end
-      #
-      #     def self.not_cool
-      #       exclude cool
-      #     end
-      #   end
-      def exclude(query)
-        query.negate!
-        query
-      end
+    # Initialize a new instance
+    #
+    # @return [Hanami::Repository] the new instance
+    #
+    # @since x.x.x
+    def initialize
+      super(self.class.container)
+    end
 
-      # This is a method to check entity persited or not
-      #
-      # @param entity
-      # @return a boolean value
-      # @since 0.3.1
-      def _persisted?(entity)
-        !!entity.id
-      end
+    # Find by primary key
+    #
+    # @return [Hanami::Entity,NilClass] the entity, if found
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   repository = UserRepository.new
+    #   user       = repository.create(name: 'Luca')
+    #
+    #   user       = repository.find(user.id)
+    def find(id)
+      root.by_primary_key(id).as(:entity).one
+    end
 
-      # Update timestamps
-      #
-      # @param entity [Object, Hanami::Entity] the entity
-      #
-      # @api private
-      # @since 0.3.1
-      def _touch(entity)
-        now = Time.now.utc
+    # Return all the records for the relation
+    #
+    # @return [Array<Hanami::Entity>] all the entities
+    #
+    # @since x.x.x
+    #
+    # @example
+    #   UserRepository.new.all
+    def all
+      root.as(:entity)
```

Comment:
I think I'd prefer to keep it chainable, in terms of being friendly to users. Don't feel super strongly though.

I could see someone having `UserRepository.new.all` then adding a where, like:`UserRepository.new.all.where(first_name: "Andy")`. The `all` is unnecessary, but it also doesn't really hurt anything either. ¯\\\_(ツ)\_/¯ 
Otherwise they'll get `NoMethodError: undefined method 'where' for []:Array`. If users want to materialize the result, they can call `to_a` themselves. It'd be surprising to me if `all` did that.